src/librbd/cache/ImageCache.h

```diff
@@ -6,21 +6,77 @@
 
 #include "include/buffer_fwd.h"
 #include "include/int_types.h"
+#include "librbd/io/Types.h"
+#include "common/ceph_json.h"
 #include <vector>
 
 class Context;
 
 namespace librbd {
+  struct ImageCtx;
 namespace cache {
 
+enum ImageCacheType {
+  IMAGE_CACHE_TYPE_RWL = 1,
+  IMAGE_CACHE_TYPE_UNKNOWN = 2,
+};
+
+template <typename ImageCtxT = ImageCtx>
+class ImageCacheState {
+protected:
+  ImageCtxT* m_image_ctx;
+
+  ImageCacheState(ImageCtxT* image_ctx): m_image_ctx(image_ctx) {}
+
+  ImageCacheState(ImageCtxT* image_ctx, JSONFormattable &f): m_image_ctx(image_ctx)  {
+    m_present = (bool)f["present"];
+    m_empty = (bool)f["empty"];
+    m_clean = (bool)f["clean"];
+  }
+  ImageCacheState(ImageCtxT* image_ctx, bool _present, bool _empty, bool _clean) :
+    m_image_ctx(image_ctx), m_present(_present), m_empty(_empty), m_clean(_clean) {}
+
+public:
+  bool m_present = false;
+  bool m_empty = true;
+  bool m_clean = true;
+
+  virtual ~ImageCacheState() {}
+
+  void write_image_cache_state(Context *on_finish);
+
+  std::string get_json_string() {
```

Comment:
Instead of two hacked together methods to build json objects, just pass a `ceph::Formatter` to be used to construct the json (or a `JSONObj`)

Reply:
Updated